pkg/services/ngalert/state/cache.go

```diff
@@ -33,6 +34,44 @@ func newCache() *cache {
 	}
 }
 
+// RegisterMetrics registers a set of Gauges in the form of collectors for the alerts in the cache.
+func (c *cache) RegisterMetrics(r prometheus.Registerer) {
+	newAlertCountByState := func(state eval.State) prometheus.GaugeFunc {
+		return prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+			Namespace:   metrics.Namespace,
+			Subsystem:   metrics.Subsystem,
+			Name:        "alerts",
+			Help:        "How many alerts by state.",
+			ConstLabels: prometheus.Labels{"state": strings.ToLower(state.String())},
+		}, func() float64 {
+			return c.countAlertsBy(state)
+		})
+	}
+
+	r.MustRegister(newAlertCountByState(eval.Normal))
+	r.MustRegister(newAlertCountByState(eval.Alerting))
+	r.MustRegister(newAlertCountByState(eval.Pending))
+	r.MustRegister(newAlertCountByState(eval.Error))
+	r.MustRegister(newAlertCountByState(eval.NoData))
+}
+
+func (c *cache) countAlertsBy(state eval.State) float64 {
+	c.mtxStates.RLock()
```

Comment:
Nit: I much prefer using a defer as then it becomes impossible to forget to unlock the mutex. Thinking about the future if or when this function is changed and needs to return errors, for example.

```suggestion
	c.mtxStates.RLock()
	defer c.mtxStates.RUnlock()
```

Reply:
In the general sense, I agree with you, but when looping over structs that have deep nests such as a maps of maps I prefer to be explicit on when do I want to hold the lock. Mostly because it can be super simple to add expensive computation in between and end up taking the lock for more than its necessary. 

That being said, I have changed it.